// urllib-next/src/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGlobalDispatcher = exports.setGlobalDispatcher = exports.MockAgent = exports.USER_AGENT = exports.HttpClient = exports.request = void 0;
const HttpClient_1 = require("./HttpClient");
let httpclient;
async function request(url, options) {
    if (!httpclient) {
        httpclient = new HttpClient_1.HttpClient();
    }
    return await httpclient.request(url, options);
}
exports.request = request;
var HttpClient_2 = require("./HttpClient");
Object.defineProperty(exports, "HttpClient", { enumerable: true, get: function () { return HttpClient_2.HttpClient; } });
Object.defineProperty(exports, "USER_AGENT", { enumerable: true, get: function () { return HttpClient_2.HEADER_USER_AGENT; } });
var undici_1 = require("undici");
Object.defineProperty(exports, "MockAgent", { enumerable: true, get: function () { return undici_1.MockAgent; } });
Object.defineProperty(exports, "setGlobalDispatcher", { enumerable: true, get: function () { return undici_1.setGlobalDispatcher; } });
Object.defineProperty(exports, "getGlobalDispatcher", { enumerable: true, get: function () { return undici_1.getGlobalDispatcher; } });
exports.default = {
    request,
    USER_AGENT: HttpClient_1.HEADER_USER_AGENT,
};
